build_tools/fish_xgettext.fish

```diff
@@ -2,84 +2,89 @@
 #
 # Tool to generate messages.pot
 
-# Create temporary directory for these operations. OS X `mktemp` is somewhat restricted, so this block
-# works around that - based on share/functions/funced.fish.
-set -q TMPDIR
-or set -l TMPDIR /tmp
-set tmpdir (mktemp -d $TMPDIR/fish.XXXXXX)
-or exit 1
+set -g output_file (status dirname)/../messages.pot
+
+# Write header. This is required by msguniq.
+# Note that this results in the file being overwritten.
+# This is desired behavior, to get rid of the results of prior invocations
+# of this script.
+begin
+    echo 'msgid ""'
+    echo 'msgstr ""'
+    echo '"Content-Type: text/plain; charset=UTF-8\n"'
+    echo ""
+end >$output_file
 
 # This is a gigantic crime.
-# xgettext still does not support rust *at all*, so we use cargo-expand to get all our wgettext invocations.
+# We use cargo-expand to get all our wgettext invocations.
+# This might be replaced once we have a tool which properly handles macro expansions.
 set -l expanded (cargo expand --lib; for f in fish fish_indent fish_key_reader; cargo expand --bin $f; end)
 
 # Extract any gettext call
 set -l strs (printf '%s\n' $expanded | grep -A1 wgettext_static_str |
              grep 'widestring::internals::core::primitive::str =' |
              string match -rg '"(.*)"' | string match -rv '^%ls$|^$' |
              # escaping difference between gettext and cargo-expand: single-quotes
-             string replace -a "\'" "'" | sort -u)
+             string replace -a "\'" "'")
 
 # Extract any constants
 set -a strs (string match -rv 'BUILD_VERSION:|PACKAGE_NAME' -- $expanded |
              string match -rg 'const [A-Z_]*: &str = "(.*)"' | string replace -a "\'" "'")
 
+# Sort the extracted strings and remove duplicates.
+# This is optional.
+set -l strs (string join \n -- $strs | sort -u)
+
 # We construct messages.pot ourselves instead of forcing this into msgmerge or whatever.
 # The escaping so far works out okay.
 for str in $strs
-    # grep -P needed for string escape to be compatible (PCRE-style),
-    # -H gives the filename.
-    # If you want to run this on non-GNU grep: Don't.
-    # The sed command extracts just the filename from the matches grep finds,
-    # and prepends the '#: ' prefix, marking the line as a source refecence.
-    # sort -u just gets rid of duplicates.
-    grep -PH -r -- \"(string escape --style=regex -- $str)\" src/ |
-       sed -E 's/^([^:]*):.*$/#: \1/' | sort -u
     echo "msgid \"$str\""
     echo 'msgstr ""'
```

Comment:
in future, maybe we can also extract a function for writing the msgid+msgstr parts (we do the same thing twice).